### ISSUE TITLE
Best point computation ?

### ISSUE THREAD

**Issue by vince62s** [rousseau-lium/XenC#3]
Hi,
I am trying to use this toolkit in a basic way with a monolingual general domain corpus + in-domain corpus. it seems to run fine but I am confused as of where the selection occurs and what exactly the best computation point is leading you to retain only 11.3% in your doc.
Do we have top manuaaly select the threshold over / under which we keep the data ?
thanks.

**Comment by fredblain** [comment]
Hi @vince62s, no need to manually fix the threshold, this is done automatically.

Please @rousseau-lium correct me if I'm wrong, but the way XenC works is: 
first, each sentence of the out-of-domain corpus is scored (LM perplexity -- ppl) against your in-domain data, and sorted accordingly. 
then, the best percentage of the out-of-domain data to keep (so-called the threshold) is determined based on a dichotomic search: a set of LMs is created successively adding 10% of the data (i.e. a LM with 10% of the data, a LM with 20% of the data, and so on up to 100%). Each LM is then assessed against the dev corpus you have to provide.
Let's say you get better ppl scores for your 10%- and 20%-based LMs. Now you know that the best percentage to keep is located between 10% and 20% of your out-of-domain data. You just need to refine it and to do so, XenC will repeat the previous process by creating new LMs but this time with a 5% step (i.e. a LM with 15% of the data, no need to do more from now on), and again with a step of successively 2% (i.e. 12%, 14%)  and 1% (i.e. 11%). 
Finally, in your case XenC returns the best point as being 11% of your out-of-domain data for which it observed the best ppl against your dev corpus.

Hope you have a better picture now.

**Comment by vince62s** [comment]
what's confusing is that it spits out 2 files xxx.mode2.sorted.gz and xxx.mode2.scored.gz but both still have 100% of the training corpus. I don't think that xxx-sample1.dev.4g.0-0-0-0.arpa is the real output LM. is it ?

**Comment by antho-rousseau** [comment]
Hi @vince62s!

Thanks @fredblain for answering on this issue! 

@vince62s, the two output files are the result of the main score computation process, i.e. they are generated before applying any threshold. They are basically the same files: one is sorted while the other one isn't. The unsorted file can be useful if you have to preserve your sentence order for any reason. 

If you only need the sorted file (which will be used for the threshold calculation process), you can pass the option --sorted-only, which will save you some time and space.
Using the -b (or --best-point) option will perform the threshold calculation process, and also the score computation process if it has not been done on a previous run.
So basically, you can do a first run, and then launch the very same command, adding -b for threshold computation, and optionally the --dev option to pass a different development corpus (by default, if no dev is specified, XenC will take the in-domain corpus as a dev), or launch everything by giving all the options in the first place.

Hope it'll answer your questions.

Best,
Anthony

**Comment by vince62s** [comment]
well I am a little stubborn ... sorry.
when we do not specify any output, is the result of the process the arpa file ?
(my command was: ./XenC --mono -s en -o train.txt -i dev.txt )
its name is xxx-sample1.deven.4g.0-0-0-0.arpa
the size of this file seems really small to me.
110 MB (not gzipped) for a training corpus of 1.5 GB

also I ran the same command with -b but after a while I get a Segmentation fault (core dumped) with no other output.

Edit : I'll retry with the -b option from the very begining.

**Comment by vince62s** [comment]
Okay, I am doing now the following experiment.
I take a 12.8 million sentence corpus of which I take out 10K sentences to be my dev set.
I build an order 4 LM on the corpus (without the dev set) and I run the ppl from the dev set: 67
If I prune my LM down to 34 MB (gzipped) the ppl becomes 168

Then I run XenC with -o my training -i my dev set
I get an arpa LM (unzipped) of 110 MB (very small)
I run the ppl with srilm I get 216

Am I doing something wrong ?

**Comment by antho-rousseau** [comment]
Hi again,
10k sentences vs. 12.8M sentences seems very small to me, but can be okay.

Your numbers appear consistent: in a case, you make a LM based on everything in your OOD corpus, which gives you a not so bad PPL; in the other case, XenC makes a LM based on a random sample from your OOD corpus, with a size as close from the dev size as possible, while never going under 1% of the OOD corpus (hence the -sample1 tag in the arpa file). It is totally logical that your resulting PPL is significantly worse. This is a normal phenomenon, it just means that the sample from your OOD is less close from your dev than the globality of your OOD.

Some things are in place in XenC to cope with such phenomenon, such as the --mean option (3 samples instead of one).

But taking a sample of your OOD to use it as the ID seems a bad idea to me (see Moore&Lewis 2010 paper "Intelligent Selection of Language Model Training Data" for instance or Amittai Axelrod's 2011 paper "Domain adaptation via pseudo in-domain data selection").

**Comment by vince62s** [comment]
"it just means that the sample from your OOD is less close from your dev than the globality of your OOD."
confusing:  my sample of OOD = dev set = ID 

but I think I see the issue I will try with more data for ID

**Comment by antho-rousseau** [comment]
Hi,
in my sentence:
- sample of OOD = sample of OOD from XenC
- dev = YOUR sample of OOD, which is still not a good idea (you want to score sentences against a subsample of them, don't see why, but why not after all)

Best.

**Comment by antho-rousseau** [comment]
BTW, having a subsample of your OOD IS a good idea as a DEV (--dev option), but should not be used as an ID corpus (of course, you may have your reasons, but I don't see the point actually).

**Comment by antho-rousseau** [comment]
Hi @vince62s, is this solved for you?

**Comment by vince62s** [comment]
Well, the thing is that I thought I could "self filer" the main corpus to see if it would optimize itself based on a dev set taken out of the main corpus. I retried with a 100k sentences but did not work better.
I have to test differently with a real ID and OOD configuration.
thanks.

**Comment by vince62s** [comment]
Hi again,
I am trying to understand exactly what the paragraph 4.3 of the paper "Enhancing the TED-LIUM Corpus with Selected Data for Language Modeling and More TED Talks".
When you say the training set is considered as the ID corpus, does this mean that in the end these data are included in the selection (ie the LM) or not ?

**Comment by antho-rousseau** [comment]
Hi @vince62s,

in the end yes, the in-domain corpus is (manually) added into the LM, but not by the tool: the tool only scores an out-of-domain corpus against an in-domain one, then is able to perform a dichotomic search of the best-point (evaluated by the perplexity on a separate dev set or the in-domain one).

Best.

**Comment by vince62s** [comment]
thanks, all clear now closing the issue.

**Comment by vince62s** [comment]
Hi back to this subject.
I tried to build the LM based on the test file given in the release2 distribution.
The LM (order 4) gives me a PPL of 144 on the dev set. I read 111 in the paper.
What could be the difference ?
I concatenated the 6 files from 6 sources + training text of acoustic data.

**Comment by vince62s** [comment]
Actually it's even more different, when I use the exact same vocabulary file:
ngram 1=152214
ngram 2=18833685 (same number as in paper : 18.8)
ngram 3=80593424 (same number as in paper : 80.6)
ngram 4=144086793 (not exactly the same number, paper : 144.2)
size is 2.4GB when paper says 2.6GB
PPL on dev set (507 sentences 17783 words) is 151.12 (paper says 111)

for doubt avoidance here is the srilm command line I used
ngram-count -text tedlium2.txt -order 4 -limit-vocab -vocab tedlium-152k.txt -unk -kndiscount -gt3min 1 -gt4min 1 -lm tedlium2.kn.gz

**Comment by vince62s** [comment]
My mistake I need to redo the merge of data sources.

So now I get the exact same number of n-grams as in the paper but the PPL is 150.57

**Comment by antho-rousseau** [comment]
Hi @vince62s,
I see at least 3 possible explanations for your difference of PPL:
- maybe not the exact same versions of the data
- probably not the same XenC version (please note that experiments are from circa 2013, paper is from 2014) => current version uses KenLM, while the paper experiments were done with the SRILM version
- there is randomness in the samples selection during the score computation

Also, paper says 113, not 111, but this is a detail.

**Comment by vince62s** [comment]
Anthony, thanks for your reply.
Maybe I was not clear enough.
From the last page of this http://www.lrec-conf.org/proceedings/lrec2014/pdf/1104_Paper.pdf
I see PPL=111 for what is called "New selection"
I downloaded the "already filtered" data that comes along with the tedlium release 2 stuff.
(it includes everything, sph, stm, LM data files...)
I just concatenated the 6 files with the training data.
I used SRILM to build the LM (cf above post the command line used)
I am getting the same number of n-grams as in the paper.
Then I run the PPL against the dev set. My number is much higher.
So in this process I did not actually used Xenc.

**Comment by antho-rousseau** [comment]
Hi @vince62s,
okay, sorry for the misunderstanding.
Actually, I'm not even sure that back in that time we released the exact same data as described in the paper... But let's assume it is the case.
One immediate difference I see is that I did not use pruning, in my experiments all cutoffs (i.e. for each order) are 0.
As the paper says:

> All language models have been estimated
> using the SRILM toolkit (Stolcke, 2002) with a modified
> Kneser-Ney discounting method and no cut-offs.

Also, you have to merge the LM data files with the text from the training set (not the dev, obviously) but I assume you did that too.

I just checked the experiments from the paper (luckily, I still have them) and: 

```
file ./TEDLIUM-dev.txt: 507 sentences, 18226 words, 0 OOVs
0 zeroprobs, logprob= -38365.9 ppl= 111.696 ppl1= 127.353
```

Here is the general command used in the oodSel LMs:

```
ngram-count -text oodSel.txt.gz -vocab 152k.vocab -order 4 -interpolate -kndiscount1 -gt1min 0 -kndiscount2 -gt2min 0 -kndiscount3 -gt3min 0 -kndiscount4 -gt4min 0 -unk -lm oodSel.152k.4g.kn-int.0-0-0-0.arpa.gz
```

You still have to interpolate that with a LM estimated on the training set text afterwards.

**Comment by vince62s** [comment]
yes I concatenated the training text with the other sources (paper says same results whether doing so or interpolating the 2 LM), and of course I did not incorporated the dev set.
I also used the no cut-off options (and I have the same number of n-grams as you which proves it's good)
The only difference I see is the -kndiscount1 2 3 4 
I only used -kndiscount maybe it makes a difference I will look into it.
Thanks for your reply
Vincent

**Comment by antho-rousseau** [comment]
Ok, hope you'll sort it out.
Best,
Anthony

**Comment by vince62s** [comment]
oh I didn't get your last comment. Are you saying that even if the training text is incorporated with the selection, you also estimated another LM with the training text and interpolated it with the LM that also includes the same data ?

**Comment by antho-rousseau** [comment]
Actually, what I did in the experiments is (but I'm not sure that's very clear in the paper):
- estimate a LM on the concatenation of the OOD data
- estimate a LM on the text from the training set
- merge those LMs together (that where the PPL 111.69 comes from)

**Comment by vince62s** [comment]
Just to follow up on this, I was able to almost reproduce the PPL (113.237)
using the above steps.
Maybe the only difference could come from the way I interpolated in the end, using mix-lm from srilm.
The paper seems to use some kind of interpolation coefficients coming from the dev set.
Anyway it's fair enough.
thanks.
